numpy/core/src/umath/ufunc_object.c

```diff
@@ -745,6 +745,29 @@ _set_out_array(PyObject *obj, PyArrayObject **store)
 
         return 0;
     }
+
+    /* If out is not an array, try calling __array__ */
+    PyObject* array_method = PyObject_GetAttrString(obj, "__array__");
```

Comment:
This approach conforms with the doc, but should we call `PyArray_FromAny` instead?